### PR TITLE
Ruby: restrict join order of API graph predicates

### PR DESCRIPTION
**Pull request by asgerf**
Places join-order restrictions on API graphs to more reliably get a good join order at the use-sites. This is done by caching more predicates, and placing inline facade predicates in front to enforce a good join ordering at the call site.

While working on adding better support for instance members in API graphs, I noticed the join order in some library models would go haywire seemingly at random, and this PR is meant to stop that from happening.

To explain, a chain of API-graph calls should be evaluated from left to right. For example,
```ql
API::getTopLevelMember("Bar").getAMethodCall("baz")
```
Should be evaluated by first finding uses of `Bar` and then calls to `baz` on those objects.

But sometimes it gets evaluated backwards, by starting at calls to `baz` and then checking if their receiver is `Bar` (it's more complex than that, but that's the basic idea). In this small example it's not so bad, but in practice backwards evaluation is really bad. Previously this was left mostly to the optimizer to figure out, which it very often did, but not always.

The PR also removes some uses of `getASubclass()` which were redundant because it came before a member predicate that itself performs `getASubclass()`. The optimizer did not handle that very well.

[Evaluation](https://github.com/github/codeql-dca-main/issues/12371) shows an average speed-up of about 2%.

### CODE REVIEW

**Comment by aschackmull** [comment]
You might get better predictability of joins by using `pragma[late_inline]` combined with a `bindingset`. That kind of join order hint is a bit more high-level and easier to reason about.

**Comment by asgerf** [comment]
@aschackmull I used `pragma[inline_late]` for `getAValueReachableFromSource` where it is a significant improvement over inlining the `only_bind` pragmas. In the other cases, I refrained from using it for a few reasons:
- For `getMethod(string m)`, it differs from call site to call site whether the parameter `m` is an input or output. Since we can only have one binding set, we'd need two variants of the predicate to support all existing use cases.
- Readability. It can't be applied directly to member predicates, which makes it a bit of a hassle to work with.
- For some predicates `inline_late` seemed to produce a _slightly_ larger tuple count than inlining `only_bind` pragmas, so given the readability issue it just didn't seem like a win.

**Comment by aschackmull** [comment]
@kaspersv FYI ^

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll [hunk 1]
```diff
@@ -85,7 +85,6 @@ module ActiveStorage {
           // Class methods
           API::getTopLevelMember("ActiveStorage")
               .getMember("Blob")
-              .getASubclass()
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll:88]: Why was this removed?

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll [hunk 2]
```diff
@@ -18,8 +18,7 @@ module Twirp {
    */
   class ServiceInstantiation extends DataFlow::CallNode {
     ServiceInstantiation() {
-      this =
-        API::getTopLevelMember("Twirp").getMember("Service").getASubclass().getAnInstantiation()
+      this = API::getTopLevelMember("Twirp").getMember("Service").getAnInstantiation()
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:21]: Here `getASubclass` has been removed, because asking for `getAnInstantiation` will include instantiations of sub classes?

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll [hunk 3]
```diff
@@ -177,15 +188,29 @@ module API {
...
+    pragma[inline]
+    Node getMember(string m) { result = pragma[only_bind_out](this).getMemberInternal(m) }
+
+    /**
+     * INTERNAL USE ONLY.
+     *
+     * Same as `getMember` but without join-order hints.
+     */
+    cached
+    Node getMemberInternal(string m) {
```
Comment (hvitved) [inline ruby/ql/lib/codeql/ruby/ApiGraphs.qll:200]: I would prefer to have such internal predicates be either top-level predicates or on a an internal `Impl` class.

**Review by hvitved — COMMENTED** [review]
LGTM, just a couple of questions.

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll (hunk 1 above)
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll:88]: From the PR description:
> The PR also removes some uses of getASubclass() which were redundant because it came before a member predicate that itself performs getASubclass(). The optimizer did not handle that very well.

File: ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll (hunk 2 above)
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/frameworks/Twirp.qll:21]: Yes, again, from the PR description:
> The PR also removes some uses of getASubclass() which were redundant because it came before a member predicate that itself performs getASubclass(). The optimizer did not handle that very well.

File: ruby/ql/lib/codeql/ruby/ApiGraphs.qll (hunk 3 above)
Comment (asgerf) [reply ruby/ql/lib/codeql/ruby/ApiGraphs.qll]: Moved to an internal class at `API::Node::Internal`.

File: ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll (hunk 1 above)
Comment (hvitved) [reply ruby/ql/lib/codeql/ruby/frameworks/ActiveStorage.qll:88]: Sorry, I missed that.

**Review by hvitved — APPROVED** [review]
